# Data Analysis with Python

<br>**Author: Tonmoy Talukder**
<br>CS Udergrad, AUST, Dhaka, Bangladesh

**Data Source:** https://stats.espncricinfo.com/ci/content/records/223646.html

**Objectives:**
> <br>Displaying the first 10 rows of the dataframe.<br>
> <br>Explaining the meaning of each column.<br>
> <br>Finding the numbers of rows and columns in the dataframe.<br>
> <br>Finding the data statistics and checking for the data types.<br>
> <br>Checking for missing values present in the dataset or not.<br>
> <br>Renaming the column names appropriately.<br>
> <br>Remove a column from the dataframe.<br>
> <br>Splitting columns.<br>

## **Importing the test_cricket.xlsx file. And reading the sheet 'wickets' from there.**

In [194]:
#importing necessary libraries 
import pandas as pd
import numpy as np

df = pd.read_excel("test_cricket.xlsx", sheet_name='wickets', engine="openpyxl")

#print("The first five rows of the data frame are:")
#display(df.head(5))

## **Displaying the first 10 rows of the dataframe**

In [195]:
# Printing first 10 rows.
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   5  
9                9/127               14/184  28.07  2.80  60.0  34   9

**Column**

In [196]:
# Printing all columns name
print(df.columns)

Index(['Player',   'Span',    'Mat',   'Inns',  'Balls',   'Runs',   'Wkts',
          'BBI',    'BBM',    'Ave',   'Econ',     'SR',        5,       10],
      dtype='object')


## **Meaning of the Each Column:** 

    1. Player: The names of the players, played test cricket.

    2. Span: The career length of the players.

    3. Mat: Number of matches played by the players.

    4. Inns: Total innings played by the players.
    
    5. Balls: The number of balls bowled. Overs is more traditional,
    but balls is a more useful statistic because the number of balls per over
    has varied historically (and even within a single match, can vary due to umpire miscounting).
    
    6. Runs: The number of runs conceded.
    
    7. Wkts: Total wickets earned by the players.
    
    8. BBI: Best bowling by the players in the innings.
    
    9. BBM: = Best bowling in the matches played by the players. 
    
    10. Ave = The average number of runs conceded per wicket taken. (Ave = Runs/W)
    
    11. Econ = The average number of runs conceded per over. (Econ = Runs/Overs bowled).
    
    12. SR = The average number of balls bowled per wicket taken. (SR = Balls/W)
    
    13. 5 = The number of innings in which the bowler took at least five wickets.
    Four wickets in an innings (4w), the number of innings in which the bowler 
    took exactly four wickets, is sometimes recorded alongside five wickets, 
    especially in limited overs cricket.
    
    14. 10 = The number of matches in which the bowler took at least ten wickets; 
    recorded for Tests and first-class matches 


## **Finding the number of rows and columns in the dataframe**

In [197]:
# Number of rows
print("Number of Rows: ")
display(df.shape[0])

# Number of columns
print("Number of Columns: ")
display(df.shape[1])

Number of Rows: 


79

Number of Columns: 


14

## **Finding the data statistics and checking for the data types**

In [198]:
# Data Statistics
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.835443  
std     3.657561   0.351666   9.349337   9.642372   3.244049  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   4.000000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

### **Observations**
- On an average a player has played 144 innings and 80 matches. <br>
- 25% of players have played 110 innings or less. <br>
- 50% of players have played 129 and 71 innings and matches or less. <br>
- 75% of players have played 169 and 93 innings and matches or less. <br>
- Maximum 301 innings and 166 matches were played by a player. <br>
- Minimum 67 innings and 37 matches were played by a player. <br>
- There are more data can be observed from this statistic. <br>

## **Checking for missing values and data types of the each columns in the dataset.**

In [199]:
# Deatil overview of the Dataset
print(df.info())

#Checking, if there are any null value in the columns.
#print('Checking if there are any null value in the columns:')
#df.isnull()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     77 non-null     object 
 8   BBM     76 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
None


### Observations:
- #### There are :

> - **3 float type variables** <br> *Bowling average, Economy rate, Strike rate.*    
> - **4 String type variables**   <br> *Best Bowling in Innings, Best Bowling in Match, Player, Span*
> - **7 integer type variables**   <br> *Match, Innings, Balls, Runs, Wickets, 5 wickets in and innings, 10 wickets in an innings*
   
- #### From the non-null columns, we can see that there are some missing values in some columns. So let's do some calculation for this.

In [200]:
# Drop a row if have at least 1 null value in any of it's columns. 
new_df = df.dropna(axis = 0, how ='any')

# Display the new dataset, without any null value of any columns. 
#display(new_df.head(5))
print("Old data set with all null rows length length:", len(df))
print("New data set with out any null row length:", len(new_df))

# Count the rows has at least 1 null values in any of it's columns. 
print("Number of rows with at least 1 Null value: ", (len(df)-len(new_df)))

# Count null values of each columns 
print("Count null values of each columns")
display(df.isnull().sum())

Old data set with all null rows length length: 79
New data set with out any null row length: 75
Number of rows with at least 1 Null value:  4
Count null values of each columns


Player    0
Span      0
Mat       0
Inns      0
Balls     0
Runs      0
Wkts      0
BBI       2
BBM       3
Ave       0
Econ      0
SR        0
5         0
10        0
dtype: int64

   >  ###### So, there arre two missing values in BBI column and three missing values in BBM colum. We can use the terminated dataset removing these two columns. And these columns are also showing unorganized datas. So, we can remove them in our later analysises.

## **Renaming columns in dataset appropriately**

In [201]:
 # Again printing columns' names to renamimng some of the columns' names
print("Columns of the Dataset: ", df.columns)

Columns of the Dataset:  Index(['Player',   'Span',    'Mat',   'Inns',  'Balls',   'Runs',   'Wkts',
          'BBI',    'BBM',    'Ave',   'Econ',     'SR',        5,       10],
      dtype='object')


In [202]:
# Renamimng the columns' names which are a little bit hard to understand
df = df.rename(columns={'Mat':'Match', 
                        'Inns':'Innings',
                        'Wkts': 'Wickets',
                        'BBI': 'Best_Bowling/Innings',
                        'BBM': 'Best_Bowling/Match',
                        'Ave': 'Average',
                        'Econ': 'Economy_Rate',
                        'SR': 'Strike_Rate',
                        5: '5_Wickets',
                        10: '10_Wickets'})

display(df.head())

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   
3        JM Anderson (ENG)  2003-2021    162      301  34791  16457      617   
4         GD McGrath (AUS)  1993-2007    124      243  29248  12186      563   

  Best_Bowling/Innings   Best_Bowling/Match  Average  Economy_Rate  \
0  1951-09-01 00:00:00               16/220    22.72          2.47   
1  1971-08-01 00:00:00               12/128    25.41          2.65   
2  1974-10-01 00:00:00               14/149    29.65          2.69   
3  1942-07-01 00:00:00  1971-11-01 00:00:00    26.67          2.83   
4  2021-08-24 00:00:00  2021-10-27 00:00:00    21.64          2.49   

   Strike_Rate  5_Wickets  10_Wickets  
0         55.0         67          22  
1         57.4         37          10  
2         65.9         35           8  
3         56.3         30           3  
4         51.9         29           3

## **Remove column from the dataframe**

In [203]:
# Remove BBI and BBM columns as they have unorganized data.
df.drop(['Best_Bowling/Innings' , 'Best_Bowling/Match'], axis=1, inplace=True)

display(df.head())

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   
3        JM Anderson (ENG)  2003-2021    162      301  34791  16457      617   
4         GD McGrath (AUS)  1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0    22.72          2.47         55.0         67          22  
1    25.41          2.65         57.4         37          10  
2    29.65          2.69         65.9         35           8  
3    26.67          2.83         56.3         30           3  
4    21.64          2.49         51.9         29           3

> *BBI and BBM columns are succesfully  removed.*

#### Split Player and Country from 'Player' Column

In [204]:
# splitting 'Player' column

df['Player'] = df['Player'].str.replace(")"," ")  
#display(df.head())
df_player = df['Player'].str.split("(", expand=True)
#display(df_player.head())

# removing Old 'Player' column
df = df.drop('Player', axis=1)  
#display(df.head())

# concatenating the new data set with the main data set
df = pd.concat([df, df_player], axis=1) 
#display(df.head())

# renaming new splitted column
df = df.rename(columns={0: 'Player',1: 'Country'})  
#display(df.head())

# new column sequence
column_sequence = ['Player' , 'Country' , 'Span' , 'Match' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets']
df = df[column_sequence]


# implementing the new column sequence in the Data Set
display(df.head())

<ipython-input-204-eab8798af341>:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['Player'] = df['Player'].str.replace(")"," ")


Player  Country       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan   ICC/SL   1992-2010    133      230  44039  18180      800   
1        SK Warne      AUS   1992-2007    145      273  40705  17995      708   
2        A Kumble    INDIA   1990-2008    132      236  40850  18355      619   
3     JM Anderson      ENG   2003-2021    162      301  34791  16457      617   
4      GD McGrath      AUS   1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0    22.72          2.47         55.0         67          22  
1    25.41          2.65         57.4         37          10  
2    29.65          2.69         65.9         35           8  
3    26.67          2.83         56.3         30           3  
4    21.64          2.49         51.9         29           3

## Players played for ICC

In [205]:
# function to check if played in ICC or not 
def ICC_Playing(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

# applying function and creating new column for playing in ICC
df['ICC_Playing'] = df['Country'].apply(ICC_Playing)

# removing "ICC/" from the country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
display(df.head(5))

Player Country       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan      SL   1992-2010    133      230  44039  18180      800   
1        SK Warne     AUS   1992-2007    145      273  40705  17995      708   
2        A Kumble   INDIA   1990-2008    132      236  40850  18355      619   
3     JM Anderson     ENG   2003-2021    162      301  34791  16457      617   
4      GD McGrath     AUS   1993-2007    124      243  29248  12186      563   

   Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets ICC_Playing  
0    22.72          2.47         55.0         67          22         Yes  
1    25.41          2.65         57.4         37          10          No  
2    29.65          2.69         65.9         35           8          No  
3    26.67          2.83         56.3         30           3          No  
4    21.64          2.49         51.9         29           3          No

In [206]:
# Count number of total ICC ICC_Playing Player
print("Number of total ICC ICC_Playing Players: ")
print(df['ICC_Playing'].value_counts())

Number of total ICC ICC_Playing Players: 
No     74
Yes     5
Name: ICC_Playing, dtype: int64


> *In total, there are only **5 ICC Playing players** in the data set, where **74 players didn't play for ICC**.*

## Number of the different countries are present in the dataset

In [207]:
print(df['Country'].value_counts())

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
PAK          7
NZ           7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64


> So, most of the players from **Austrarila (18)**, 2nd and 3rd most are following **Engalnd (13)** and **India (10)**. And **others countries have less than 10 player(s)**.

   > ### Number of *Australian Bowlers* are present in the dataset: *18*

   > ### Number of *Bangladeshi player* is present in the dataset: *1*

#### Splitting span as Start Year and Last Year

In [208]:
# splitting 'Carrer_Duration' column
df_carrer_duration = df['Span'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_carrer_duration], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Start_Year",
                        1: "End_Year"})

# removing 'Carrer_Duration' column
df = df.drop("Span", axis=1)

# new column sequence as a list
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Match' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets', 'ICC_Playing']

# implementing the new column sequence
df = df[column_sequence]  

#### Changing Start and End Year from object type to integer

In [209]:
df['End_Year']=pd.to_numeric(df['End_Year'])
df['Start_Year']=pd.to_numeric(df['Start_Year'])

In [210]:
# total years of career of the players
df['Career_Span']=df['End_Year']-df['Start_Year']

# rearranging columns
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Career_Span' , 'Match', 'Innings', 'Balls', 'Runs', 'Wickets' , 'Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets', 'ICC_Playing']
df = df[column_sequence] 
display(df.head())

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
0  M Muralitharan      SL         1992      2010           18    133      230   
1        SK Warne     AUS         1992      2007           15    145      273   
2        A Kumble   INDIA         1990      2008           18    132      236   
3     JM Anderson     ENG         2003      2021           18    162      301   
4      GD McGrath     AUS         1993      2007           14    124      243   

   Balls   Runs  Wickets  Average  Economy_Rate  Strike_Rate  5_Wickets  \
0  44039  18180      800    22.72          2.47         55.0         67   
1  40705  17995      708    25.41          2.65         57.4         37   
2  40850  18355      619    29.65          2.69         65.9         35   
3  34791  16457      617    26.67          2.83         56.3         30   
4  29248  12186      563    21.64          2.49         51.9         29   

   10_Wickets ICC_Playing  
0          22         Yes  
1          10          No  
2           8          No  
3           3          No  
4           3          No

## Player(s) had played for the longest period of time

In [153]:
display(df.sort_values(by='Career_Span', ascending = False).head(3))

Player Country  Match  Innings  Balls   Runs  Wickets  Average  \
21    Imran Khan     PAK      88      142  19458   8258      362    22.81   
55     GS Sobers      WI      93      159  21599   7999      235    34.03   
9   HMRKB Herath      SL      93      170  25993  12157      433    28.07   

    Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  Start_Year  End_Year  \
21          2.54         53.7         23           6        1971      1992   
55          2.22         91.9          6           0        1954      1974   
9           2.80         60.0         34           9        1999      2018   

    Career_Span  
21           21  
55           20  
9            19

> Imran Khan has longest carrer span of 21 years. And 2nd and third longest are following by GS Sobers (20) and HMRKB Herath (19)

## Player(s) had played for the shortest period of time

In [154]:
display(df.sort_values(by='Career_Span', ascending = True).head(3))

Player Country  Match  Innings  Balls  Runs  Wickets  Average  \
44    GP Swann     ENG      60      109  15349  7642      255    29.96   
71    K Rabada      SA      47       86   8785  4846      213    22.75   
54  Yasir Shah     PAK      45       83  13601  7247      235    30.83   

    Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  Start_Year  End_Year  \
44          2.98         60.1         17           3        2008      2013   
71          3.30         41.2         10           4        2015      2021   
54          3.19         57.8         16           3        2014      2021   

    Career_Span  
44            5  
71            6  
54            7

> GP Swann Khan has shortest carrer span of 5 years. And 2nd and third shortest are following by K Rabada (6) and Yasir Shah (7)

## The player had the lowest economy rate

In [155]:
display(df.sort_values(by='Economy_Rate', ascending = True).head(1))

Player Country  Match  Innings  Balls  Runs  Wickets  Average  \
32  LR Gibbs      WI      79      148  27115  8989      309    29.09   

    Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  Start_Year  End_Year  \
32          1.98         87.7         18           2        1958      1976   

    Career_Span  
32           18

> LR Gibbs has the lowest economy rate of 1.98

## The player had the lowest strike rate

In [156]:
display(df.sort_values(by='Strike_Rate', ascending = True).head(1))

Player Country  Match  Innings  Balls  Runs  Wickets  Average  \
71  K Rabada      SA      47       86   8785  4846      213    22.75   

    Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  Start_Year  End_Year  \
71           3.3         41.2         10           4        2015      2021   

    Career_Span  
71            6

> K Rabada has the lowest strike rate of 41.2

## The player had the lowest bowling average

In [158]:
display(df.sort_values(by='Average', ascending = True).head(1))

Player Country  Match  Innings  Balls  Runs  Wickets  Average  \
19  MD Marshall      WI      81      151  17584  7876      376    20.94   

    Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  Start_Year  End_Year  \
19          2.68         46.7         22           4        1978      1991   

    Career_Span  
19           13

> MD Marshall has the lowest bowling average of 20.94